src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/StructMarshallingStubMapNode.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+using Internal.NativeFormat;
+using Internal.Text;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents a hash table of struct marshalling stub types generated into the image.
+    /// </summary>
+    internal sealed class StructMarshallingStubMapNode : ObjectNode, ISymbolDefinitionNode
+    {
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        private ExternalReferencesTableNode _externalReferences;
+
+        public StructMarshallingStubMapNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__struct_marshalling_stub_map_End", true);
+            _externalReferences = externalReferences;
+        }
+
+        public ISymbolDefinitionNode EndSymbol => _endSymbol;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__struct_marshalling_stub_map");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => _externalReferences.Section;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName(NodeFactory factory) => this.GetMangledName(factory.NameMangler);
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node does not trigger generation of other nodes.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolDefinitionNode[] { this });
+
+            var writer = new NativeWriter();
+            var typeMapHashTable = new VertexHashtable();
+
+            Section hashTableSection = writer.NewSection();
+            hashTableSection.Place(typeMapHashTable);
+
+            foreach (var structEntry in factory.InteropStubManager.GetStructMarshallingTypes())
+            {
+
+                // the order of data written is as follows:
+                //  0. managed struct type
+                //  1. native struct type
+                //  2. struct marshalling thunk
+                //  3. struct unmarshalling thunk
+                //  4. struct cleanup thunk
+                //  5.  NumFields << 1 | HasInvalidLayout 
+                //  6  for each field
```

Comment:
This looks awfully similar to what's available in reflection metadata. At some point (definitely not now) we should reconcile this and only emit it for non-reflectable fields.